h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -522,76 +521,137 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
 
         // extract predictors
         int nonZeroCount = 0;
-        int currentRow = 0;
+        int rowPointer = 0;
         int currentCol = 0;
         int rwRow = 0;
+        
+        int rowHeaderRowPointer = 0;
+        int rowHeaderColPointer = 0;
+        
+        int lastNonZeroRow = 0;
 
         for (Integer chunk : chunks) {
             for(long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk+1]; i++) {
                 if (w != null && w.at(i) == 0) continue;
-
-                final int startNonZeroCount = nonZeroCount;
-                // enlarge final data arrays by 2x if needed
+                actualRows++;
+                if(rowHeaderColPointer == SPARSE_MATRIX_DIM){
+                    rowHeaderColPointer = 0;
+                    rowHeaderRowPointer++;
+                }
+                boolean foundNonZero = false;
 
                 for (int j = 0; j < di._cats; ++j) {
-                    data[currentRow][currentCol] = 1; //one-hot encoding
+                    if(currentCol == SPARSE_MATRIX_DIM){
+                        currentCol = 0;
+                        rowPointer++;
+                    }
+                    data[rowPointer][currentCol] = 1; //one-hot encoding
+                    
+                    if(!foundNonZero){
+                        foundNonZero = true;
+                        for (int k = lastNonZeroRow; k < actualRows; k++) {
+                        rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;
```

Comment:
This doesn't seem correct to me. IMHO every time you increment a pointer you need to make sure there is room in the array.

Is there a mechanism that prevents you from running out of space in `rowHeaders[rowHeaderRowPointer]`

Reply:
There is a prevention mechanism. It is actually so simple it is easy to overlook. 

This particular "row header" (as we call it) has one purpose: to mark the first non-zero element each line begins with. As it required for this to happen exactly once per each row (if a row is empty, last non-zero is re-used, there is this special handling a little bit further in the code), the increment only happens once. Therefore, there is no need to check it more than once per each line.

This implies the actual check for `rowHeaderColPointer` overflow is NOT done in the ` for (int j = 0; j < di._cats; ++j) ` loop, but rather directly in the embedding loop `for(long i = f.anyVec().espc()[chunk]; i < f.anyVec().espc()[chunk+1]; i++)`. It is done right after weights are checked not to be null and the line verified for not to be skipped.

This piece of code does it:
```java
                if(rowHeaderColPointer == SPARSE_MATRIX_DIM){
                // enlarge final data arrays by 2x if needed
                    rowHeaderColPointer = 0;
                    rowHeaderRowPointer++;
                }
                boolean foundNonZero = false;
```

As seen in the example, there is also `foundNonZero` variable defined and explicitely set to false. This variable makes it possible to find out if there was a non-zero element found already on this row (as the outer loop is naturally traversing the matrix row-wise).

And as you can see a little bit further on line 550:

```java
                    if(!foundNonZero){
                        foundNonZero = true;
....
```

The non-zero element is only written into row headers if there was none discovered on this line before.

Reply:
Thanks!